### PR TITLE
[ws client] RegisterNotification support

### LINKED ISSUE #301: [ws client] How to handle incoming notifications?
I'm using jsonrpsee to implement a websocket rpc client, and I need to handle some incoming notifications.  I don't see a way to register a handler for incoming notifications and curious if I'm missing something in the docs / examples.  Is this not supported?

Thanks!

### PR DESCRIPTION
**Pull request by billylindeman**
This adds the ability to register a NotificationHandler\<T\> for a method name, allowing jsonrpsee ws clients to support server originated notifications!  #301 

Usage:

```
let mut offers: NotificationHandler<SessionDescription> = client.register_notification("offer").await?;
```

This is my first PR on jsonrpsee so let me know what you think :). I think the error cases could use a second glance by someone more experienced on the project but it does work ATM.

@niklasad1 @dvdplm 

fixes #301

### CODE REVIEW

File: ws-client/src/helpers.rs [hunk 1]
```diff
@@ -72,6 +72,33 @@ pub fn process_subscription_response(
...
+		Some(send_back_sink) => match send_back_sink.try_send(notif.params) {
+			Ok(()) => Ok(()),
+			Err(err) => {
+				log::error!("Error sending notification subscription {:?} error: {:?}", sub_id, err);
+				Err(Error::Internal(err.into_send_error()))
+			}
+		},
+		None => {
+			log::error!("Subscription ID: {:?} not an active subscription", sub_id);
+			Err(Error::InvalidSubscriptionId)
```
Comment (billylindeman) [inline ws-client/src/helpers.rs:97]: This error is a bit confusing on the frontend so maybe we should have a specific error for unexpected notification

**Review by niklasad1 — COMMENTED** [review]
Looks overall good, thanks.

However, I think the code could be simplified/more readable by just using the method name and the Request Manager but it might needs some additional boiler plate code I suppose.

File: types/src/client.rs [hunk 2]
```diff
@@ -56,6 +56,19 @@ pub struct SubscriptionMessage {
 	pub send_back: oneshot::Sender<Result<(mpsc::Receiver<JsonValue>, SubscriptionId), Error>>,
 }
 
+/// OnNotification message.
+#[derive(Debug)]
+pub struct OnNotificationMessage {
+	/// Request ID of the subscribe message.
+	pub req_id: u64,
+	/// SubscriptionId the method name this notification handler is attached to
+	pub sub_id: SubscriptionId,
```
Comment (niklasad1) [inline types/src/client.rs:65]:
```suggestion
	pub method_name: SubscriptionId,
```

File: types/src/client.rs [hunk 3]
```diff
@@ -56,6 +56,19 @@ pub struct SubscriptionMessage {
 	pub send_back: oneshot::Sender<Result<(mpsc::Receiver<JsonValue>, SubscriptionId), Error>>,
 }
 
+/// OnNotification message.
+#[derive(Debug)]
+pub struct OnNotificationMessage {
+	/// Request ID of the subscribe message.
+	pub req_id: u64,
```
Comment (niklasad1) [inline types/src/client.rs:63]: This seems needless given that we only care about that `method_name`

I assume this is a workaround to work to work how current subscriptions are handled in the `RequestManager`.

File: types/src/traits.rs [hunk 4]
```diff
@@ -44,6 +44,11 @@ pub trait SubscriptionClient: Client {
 	) -> Result<Subscription<Notif>, Error>
 	where
 		Notif: DeserializeOwned;
+
+	/// Register a Subscription<Notif> that will listen for incoming JSON-RPC notifications
+	async fn on_notification<'a, Notif>(&self, method: &'a str) -> Result<Subscription<Notif>, Error>
```
Comment (niklasad1) [inline types/src/traits.rs:49]: `register_notification` seems like a more appropriate name for this?

File: ws-client/src/client.rs [hunk 5]
```diff
@@ -550,6 +587,20 @@ async fn background_task(
...
 				}
 			}
+
+			// User called `on_notification` on the front-end.
+			Either::Left((Some(FrontToBack::OnNotification(sub)), _)) => {
+				log::trace!("[backend] registering notification handler: {:?}", sub.sub_id);
+				let (subscribe_tx, subscribe_rx) = mpsc::channel(max_notifs_per_subscription);
+
+				if manager.insert_notification_handler(sub.req_id, sub.sub_id.clone(), subscribe_tx).is_ok() {
+					sub.send_back
```
Comment (niklasad1) [inline ws-client/src/client.rs:597]: This expect should be removed because if the user drops the notification handle to this will panic,

so just ignore the error here by `let _ = send(...)`

File: ws-client/src/helpers.rs (hunk 1 above)
Comment (niklasad1) [reply ws-client/src/helpers.rs:97]: Yeah, just add a new error `InvalidNotification`, `UnregisteredNotification` or similar.

File: ws-client/src/client.rs [hunk 6]
```diff
@@ -455,6 +456,42 @@ impl SubscriptionClient for WsClient {
 		};
 		Ok(Subscription { to_back: self.to_back.clone(), notifs_rx, marker: PhantomData, id })
 	}
+
+	/// Register a notification handler for async messages from the server.
+	///
+	async fn on_notification<'a, N>(&self, method: &'a str) -> Result<Subscription<N>, Error>
```
Comment (niklasad1) [inline ws-client/src/client.rs:462]: I think we need a new type for `Subscription`, the drop implementation will not work I think.

Comment (billylindeman) [reply ws-client/src/client.rs:462]: Yeah agreed actually this would let me clean things up quite a bit. I'll work on that today

File: types/src/client.rs (hunk 3 above)
Comment (billylindeman) [reply types/src/client.rs:63]: Yep, I will refactor out

File: ws-client/src/client.rs (hunk 5 above)
Comment (billylindeman) [reply ws-client/src/client.rs]: Are you sure about this one?  This insert should happen in the code path for register_notification() and that async is blocked on this send (or it should fail to register).

File: types/src/client.rs [hunk 7]
```diff
@@ -132,3 +134,13 @@ impl<Notif> Drop for Subscription<Notif> {
 		let _ = self.to_back.send(FrontToBack::SubscriptionClosed(id)).now_or_never();
 	}
 }
+
+impl<Notif> Drop for NotificationHandler<Notif> {
```
Comment (billylindeman) [inline types/src/client.rs:138]: @niklasad1 Can you take a look at my drop implementation / follow this code path?

Comment (niklasad1) [reply types/src/client.rs:138]: it looks good

File: types/src/client.rs [hunk 8]
```diff
@@ -132,3 +134,13 @@ impl<Notif> Drop for Subscription<Notif> {
...
+	fn drop(&mut self) {
+		// We can't actually guarantee that this goes through. If the background task is busy, then
```
Comment (niklasad1) [inline types/src/client.rs:140]: can you update this comment? little outdated now.

File: types/src/client.rs [hunk 9]
```diff
@@ -132,3 +134,13 @@ impl<Notif> Drop for Subscription<Notif> {
...
+		// the channel's buffer will be full, and our unsubscription request will never make it.
+		// However, when a notification arrives, the background task will realize that the channel
+		// to the `Subscription` has been closed, and will perform the unsubscribe.
+		let _ = self.to_back.send(FrontToBack::UnregisterNotification((&self.method).to_owned())).now_or_never();
```
Comment (niklasad1) [inline types/src/client.rs:144]:
```suggestion
        let notif_method = std::mem::replace(self.method);
		let _ = self.to_back.send(FrontToBack::UnregisterNotification(notif_method)).now_or_never();
```

File: ws-client/src/client.rs [hunk 10]
```diff
@@ -602,6 +602,12 @@ async fn background_task(
 					let _ = reg.send_back.send(Err(Error::MethodAlreadyRegistered(reg.method)));
 				}
 			}
+
+			// User called `on_notification` on the front-end.
+			Either::Left((Some(FrontToBack::UnregisterNotification(method)), _)) => {
+				log::trace!("[backend] unregistering notification handler: {:?}", method);
+				let _ = manager.remove_notification_handler(&method);
```
Comment (niklasad1) [inline ws-client/src/client.rs:609]: I think you can pass this by value instead then you don't have to clone/allocate in `remove_notification`

File: ws-client/src/client.rs (hunk 5 above)
Comment (niklasad1) [reply ws-client/src/client.rs]: That's a good comment but the reason is that the user may drop the `register_notifiction future` which is possible and then this will panic AFAIU ^^

Probably quite unlikely but better be safe than sorry I guess

File: ws-client/src/client.rs [hunk 11]
```diff
@@ -455,6 +456,43 @@ impl SubscriptionClient for WsClient {
...
+			Ok(Ok(val)) => val,
+			Ok(Err(err)) => return Err(err),
+			Err(_) => return Err(self.read_error_from_backend().await),
+		};
+
+		Ok(NotificationHandler {
+			to_back: self.to_back.clone(),
+			notifs_rx,
+			marker: PhantomData,
+			method: method.to_owned(),
```
Comment (niklasad1) [inline ws-client/src/client.rs:493]:
```suggestion
			method,
```

File: ws-client/src/manager.rs [hunk 12]
```diff
@@ -240,6 +261,17 @@ impl RequestManager {
 		}
 	}
 
+	/// Get a mutable reference to underlying `Sink` in order to send incmoing notifications to the subscription.
+	///
+	/// Returns `Some` if the `request_id` was registered as a NotificationHandler otherwise `None`.
+	pub fn as_notification_handler_mut(&mut self, method: &str) -> Option<&mut SubscriptionSink> {
+		if let Some(sink) = self.notification_handlers.get_mut(&method.to_owned()) {
```
Comment (niklasad1) [inline ws-client/src/manager.rs:268]:
```suggestion
		self.notification_handlers.get_mut(method)
```

File: ws-client/src/client.rs (hunk 5 above)
Comment (billylindeman) [reply ws-client/src/client.rs]: 👍  ah that does make sense

File: ws-client/src/helpers.rs [hunk 13]
```diff
@@ -72,6 +72,27 @@ pub fn process_subscription_response(
...
+///
+/// Returns Ok() if the response was successfully handled
+/// Returns Err() if there was no handler for the method
+
+pub fn process_notification(manager: &mut RequestManager, notif: JsonRpcNotifResponse<JsonValue>) -> Result<(), Error> {
+	match manager.as_notification_handler_mut(&notif.method.to_owned()) {
+		Some(send_back_sink) => match send_back_sink.try_send(notif.params) {
+			Ok(()) => Ok(()),
+			Err(err) => {
+				log::error!("Error sending notification subscription {:?} error: {:?}", notif.method, err);
```
Comment (niklasad1) [inline ws-client/src/helpers.rs:85]: An error here will be that the channel is full which means that the user couldn't keep with the notifications.

We should drop the notification to indicate this to the user, we do that for subscriptions.

**Comment by billylindeman** [comment]
Thanks for all the feed back @niklasad1 I should be able to address all these tomorrow and update the PR

File: types/src/client.rs [hunk 14]
```diff
@@ -17,6 +17,18 @@ pub struct Subscription<Notif> {
 	pub marker: PhantomData<Notif>,
 }
 
+/// Active NotificationHandler on a Client.
+pub struct NotificationHandler<Notif> {
+	/// Channel to send requests to the background task.
+	pub to_back: mpsc::Sender<FrontToBack>,
+	/// Channel from which we receive notifications from the server, as encoded `JsonValue`s.
+	pub notifs_rx: mpsc::Receiver<JsonValue>,
+	/// Method,
```
Comment (dvdplm) [inline types/src/client.rs:26]:
```suggestion
	/// Method name
```
Maybe it'll be clearer once I read through, but why is this needed?

File: types/src/client.rs [hunk 15]
```diff
@@ -17,6 +17,18 @@ pub struct Subscription<Notif> {
...
+	pub method: String,
+	/// Marker in order to pin the `Notif` parameter.
+	pub marker: PhantomData<Notif>,
```
Comment (dvdplm) [inline types/src/client.rs:29]: Does this need to be `pub`?

File: types/src/error.rs [hunk 16]
```diff
@@ -67,6 +67,9 @@ pub enum Error {
 	/// Invalid request ID.
 	#[error("Invalid request ID")]
 	InvalidRequestId,
+	/// Unregistered notification method
```
Comment (dvdplm) [inline types/src/error.rs:70]: Can we extend the docs here a little bit, it's not immediately clear (to me) what "notification" means here, it being a word used for a few different things in the library. Maybe just a note about when this can happen?

File: ws-client/src/client.rs [hunk 17]
```diff
@@ -455,6 +456,43 @@ impl SubscriptionClient for WsClient {
...
+
+		let res = send_back_rx.await;
+		let (notifs_rx, method) = match res {
+			Ok(Ok(val)) => val,
+			Ok(Err(err)) => return Err(err),
+			Err(_) => return Err(self.read_error_from_backend().await),
+		};
+
+		Ok(NotificationHandler {
+			to_back: self.to_back.clone(),
```
Comment (dvdplm) [inline ws-client/src/client.rs:490]: Do you really need to clone `to_back` both here and on line 470 above?

File: ws-client/src/client.rs [hunk 18]
```diff
@@ -550,6 +588,26 @@ async fn background_task(
...
+				if manager.insert_notification_handler(&reg.method, subscribe_tx).is_ok() {
+					reg.send_back
+						.send(Ok((subscribe_rx, reg.method)))
+						.expect("error sending response for notification handler");
+				} else {
+					let _ = reg.send_back.send(Err(Error::MethodAlreadyRegistered(reg.method)));
+				}
+			}
+
+			// User called `on_notification` on the front-end.
```
Comment (dvdplm) [inline ws-client/src/client.rs:606]: Is this comment correct? Why does calling `on_notification` cause de-registration?

File: ws-client/src/client.rs [hunk 19]
```diff
@@ -566,12 +624,23 @@ async fn background_task(
...
 					log::debug!("[backend]: recv subscription {:?}", notif);
 					if let Err(Some(unsub)) = process_subscription_response(&mut manager, notif) {
 						let _ = stop_subscription(&mut sender, &mut manager, unsub).await;
 					}
 				}
+				// Incoming Notification
+				else if let Ok(notif) = serde_json::from_slice::<JsonRpcNotifResponse<_>>(&raw) {
+					log::debug!("[backend]: recv notification {:?}", notif);
+					match process_notification(&mut manager, notif) {
+						Ok(_) => return,
```
Comment (dvdplm) [inline ws-client/src/client.rs:637]: @niklasad1 is it correct to exit the `loop` here?

**Review by dvdplm — COMMENTED** [review]
Good progress here, tyvm.

I for one could really use a proper example of how to use this, it would help clarify what the API looks like. Do you mind adding one?

File: types/src/client.rs [hunk 20]
```diff
@@ -17,6 +17,18 @@ pub struct Subscription<Notif> {
 	pub marker: PhantomData<Notif>,
 }
 
+/// Active NotificationHandler on a Client.
+pub struct NotificationHandler<Notif> {
```
[the remark this answers is not in the document]
Comment (niklasad1) [reply types/src/client.rs:21]: No, the issue is the drop implementation but I guess we could have a flag or something internally.

This is fine for me though :)

File: ws-client/src/manager.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (niklasad1) [reply ws-client/src/manager.rs:268]: `get_mut` should be enough ^

File: types/src/client.rs (hunk 14 above)
Comment (niklasad1) [reply types/src/client.rs:26]: It's need because the user declares that wants to "subscribe" for all notifications for `method` then this is used to filter out notifications that the user it interested in.

It's in essence similar to `subscriptions` but then the subscription ID is used for that instead.

File: ws-client/src/client.rs (hunk 18 above)
Comment (niklasad1) [inline ws-client/src/client.rs:606]:
```suggestion
			// User called `register_notification` on the front-end.
```

File: ws-client/src/client.rs (hunk 19 above)
Comment (niklasad1) [reply ws-client/src/client.rs:637]: nice catch, it shouldn't

File: types/src/client.rs (hunk 20 above)
[the remark this answers is not in the document]
Comment (dvdplm) [reply types/src/client.rs:21]: I tried this and it compiles but I realized there are no tests so my code is likely wrong then. 

My drop impl just became `let _ = self.inner.to_back.send(FrontToBack::UnregisterNotification((&self.method).to_owned())).now_or_never();`

**Review by dvdplm — CHANGES_REQUESTED** [review]
Some good tests would be good.

File: types/src/client.rs (hunk 20 above)
[the remark this answers is not in the document]
Comment (niklasad1) [reply types/src/client.rs:21]: yeah, that should work too.

File: types/src/client.rs (hunk 15 above)
Comment (billylindeman) [reply types/src/client.rs:29]: Not sure.. @niklasad1?  It is pub in the Subscription<T>

Comment (niklasad1) [reply types/src/client.rs:29]: It needs to be pub unless we provide a `fn new` ^

File: ws-client/src/client.rs [hunk 21]
```diff
@@ -595,18 +590,16 @@ async fn background_task(
...
-						.send(Ok((subscribe_rx, reg.method)))
-						.expect("error sending response for notification handler");
+					let _ = reg.send_back.send(Ok((subscribe_rx, reg.method)));
 				} else {
 					let _ = reg.send_back.send(Err(Error::MethodAlreadyRegistered(reg.method)));
 				}
 			}
 
-			// User called `on_notification` on the front-end.
+			// User dopped the notificationHandler for this method
```
[the remark this answers is not in the document]
Comment (billylindeman) [reply ws-client/src/client.rs:599]: yes  /sigh

File: ws-client/src/client.rs [hunk 22]
```diff
@@ -634,10 +627,10 @@ async fn background_task(
 				else if let Ok(notif) = serde_json::from_slice::<JsonRpcNotifResponse<_>>(&raw) {
 					log::debug!("[backend]: recv notification {:?}", notif);
 					match process_notification(&mut manager, notif) {
-						Ok(_) => return,
+						Ok(_) => {}
 						Err(err) => {
 							let _ = front_error.send(err);
-							return;
+							break;
```
[the remark this answers is not in the document]
Comment (niklasad1) [reply ws-client/src/client.rs:633]: It doesn't matter in this context, so I was just inconsistent I guess ^

File: ws-client/src/manager.rs [hunk 23]
```diff
@@ -240,6 +261,13 @@ impl RequestManager {
 		}
 	}
 
+	/// Get a mutable reference to underlying `Sink` in order to send incmoing notifications to the subscription.
+	///
+	/// Returns `Some` if the `request_id` was registered as a NotificationHandler otherwise `None`.
+	pub fn as_notification_handler_mut(&mut self, method: &str) -> Option<&mut SubscriptionSink> {
+		self.notification_handlers.get_mut(&method.to_owned())
```
Comment (niklasad1) [inline ws-client/src/manager.rs:268]:
```suggestion
		self.notification_handlers.get_mut(method)
```

File: ws-client/src/helpers.rs (hunk 13 above)
Comment (niklasad1) [reply ws-client/src/helpers.rs:84]: Then the user has re-register the method again to use it again or increase the size of buffer ^

File: types/src/client.rs (hunk 20 above)
[the remark this answers is not in the document]
Comment (billylindeman) [reply types/src/client.rs:21]: > Some good tests would be good.

@dvdplm I added a basic test case for the happy path.  Open to suggestions for more tests

**Comment by billylindeman** [comment]
@dvdplm I added notification support to the mock server and a test for it.  Any other test suggestions?

File: ws-client/src/helpers.rs [hunk 24]
```diff
@@ -72,6 +72,27 @@ pub fn process_subscription_response(
...
+///
+/// Returns Ok() if the response was successfully handled
+/// Returns Err() if there was no handler for the method
+pub fn process_notification(manager: &mut RequestManager, notif: JsonRpcNotifResponse<JsonValue>) -> Result<(), Error> {
+	match manager.as_notification_handler_mut(notif.method.to_owned()) {
+		Some(send_back_sink) => match send_back_sink.try_send(notif.params) {
+			Ok(()) => Ok(()),
+			Err(err) => {
+				log::error!("Error sending notification, dropping handler for {:?} error: {:?}", notif.method, err);
+				let _ = manager.remove_notification_handler(notif.method.to_owned());
```
Comment (billylindeman) [inline ws-client/src/helpers.rs:85]: @niklasad1 So we should remove the handler her / drop it in this scope?

Comment (niklasad1) [reply ws-client/src/helpers.rs:85]: yepp, that should be it.

if you could just port this [test](https://github.com/paritytech/jsonrpsee/blob/master/tests/tests/integration_tests.rs#L139-#L166) to the register_notification that would be great, then it should be good

feel free to split it into two tests if you want.

File: test-utils/src/types.rs [hunk 25]
```diff
@@ -114,6 +116,16 @@ impl WebSocketTestServer {
 		Self { local_addr, exit: tx }
 	}
 
+	// Spawns a dummy `JSONRPC v2` WebSocket server that sends out a pre-configured `hardcoded notification` for every connection.
+	pub async fn with_hardcoded_notification(sockaddr: SocketAddr, notification: String) -> Self {
+		let listener = async_std::net::TcpListener::bind(sockaddr).await.unwrap();
```
[the remark this answers is not in the document]
Comment (niklasad1) [reply test-utils/src/types.rs:121]: It's `async-std` when I wrote it for a long time ago to avoid using `compat` stuff but we could get rid of that the make the dependency tree a little smaller

**Review by niklasad1 — COMMENTED** [review]
Thanks grinding this out,

One last thing:
Can you could just port this [test](https://github.com/paritytech/jsonrpsee/blob/master/tests/tests/integration_tests.rs#L139-#L166) to register_notification that would be great, then it should be good

File: ws-client/src/helpers.rs (hunk 24 above)
Comment (billylindeman) [reply ws-client/src/helpers.rs:85]: done

File: test-utils/src/types.rs [hunk 26]
```diff
@@ -114,6 +117,25 @@ impl WebSocketTestServer {
 		Self { local_addr, exit: tx }
 	}
 
+	// Spawns a dummy `JSONRPC v2` WebSocket server that sends out a pre-configured `hardcoded notification` for every connection.
+	pub async fn with_hardcoded_notification(sockaddr: SocketAddr, notification: String) -> Self {
+		let (tx, rx) = mpsc::channel::<()>(1);
+		let (addr_tx, addr_rx) = oneshot::channel();
+
+		std::thread::spawn(move || {
```
Comment (niklasad1) [inline test-utils/src/types.rs:125]: why spawning a thread for this instead `tokio spawn`?
